src/TT/Model.hs

```diff
@@ -27,14 +27,130 @@ data D v
   = Pi (DT v) (D v → D v)
   | Lam (D v → D v)
   | App (D v) (D v)
+  | Sg (DT v) (D v → D v)
+  | Pair (D v) (D v)
+  | Fst (D v)
+  | Snd (D v)
   | Sing (DT v) (D v)
+  | Squash (DT v)
   | Unit
   | Void
   | Abort (DT v) (D v)
   | Ax
+  | Box (D v)
   | Univ
+  | Equal (DT v) (DT v) (D v) (D v)
   | FV v
+  | Coe (DT v) (DT v) (D v) (D v)
+  | Coh (DT v) (DT v) (D v) (D v)
 
+-- | Perform application if possible
+--
+doApp
+  ∷ D v
+  → D v
+  → D v
+doApp m n =
+  case m of
+    Lam f → f m
+    Box m' → Box $ doApp m' n
+    _ → App m n
+
+-- | Project the first component of a pair if possible
+--
+doFst
+  ∷ D v
+  → D v
+doFst = \case
+  Pair m _ → m
+  Box m → Box $ doFst m
+  m → Fst m
+
+-- | Project the second component of a pair if possible
+--
+doSnd
+  ∷ D v
+  → D v
+doSnd = \case
+  Pair _ m → m
+  Box m → Box $ doSnd m
+  m → Snd m
+
+-- | Wrap a value in a box
+--
+doBox
+  ∷ D v
+  → D v
+doBox = \case
+  Box m → Box m
+  m → Box m
```

Comment:
I wonder if this is safe! Seems slightly wrong in the presence of iterated squashing, but perhaps not?